Add possibility to specify property types for generated properties

allansun opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | yes
| QA            | yes

### Description

Fixes #122

RFC: https://discourse.laminas.dev/t/rfc-laminas-laminas-code-add-type-hinting-option-in-propertygenerator/2652

allansun committed:
Add Type Hinting to PropertyGenerator
Signed-off-by: Allan Sun <allan.sun@bricre.com>

allansun committed:
Run composer cs-fix

Signed-off-by: Allan Sun <allan.sun@bricre.com>

samsonasik commented on .gitignore:
```diff
@@ -6,3 +6,4 @@
 /laminas-mkdoc-theme/
 /phpunit.xml
 /vendor/
+/.idea/
```
please revert this for consistency, you can register your own in your system's global .gitignore.

allansun committed:
Fix unit test

Signed-off-by: Allan Sun <allan.sun@bricre.com>

allansun committed:
Remove unnecessary .gitignore for entry for code consistency

Signed-off-by: Allan Sun <allan.sun@bricre.com>

samsonasik commented on src/Generator/PropertyGenerator.php:
```diff
@@ -304,4 +329,17 @@ public function omitDefaultValue(bool $omit = true)
+
+    public function setType(?string $type): void
+    {
+        if (class_exists($type) && ! str_starts_with($type, '\\')) {
```
`str_starts_with` not exists yet in php 7.4

allansun replied:
Thanks for quickly reviewing the code. 

symfony/polyfill-php80 package was included because of vimeo/psalm, so I presume it should be fine?

https://github.com/laminas/laminas-code/blob/c99ef8e5629c33bfaa3a8f1df773e916af564cd6/composer.json#L21

allansun replied:
I've updated the code to be compatible with PHP 7.4

weierophinney replied:
> symfony/polyfill-php80 package was included because of vimeo/psalm, so I presume it should be fine?

Yes, but only if you were to add it as an explicit package requirement (psalm won't be installed in production installs...)

But the changes to make it BC work, too.

allansun committed:
Run composer cs-fix
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

allansun committed:
Fix PHP 7.4 compatability
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

samsonasik commented on test/Generator/PropertyGeneratorTest.php:
```diff
@@ -387,14 +388,14 @@ public function testFromReflectionOmitsDefaultValueIfItIsNull(): void
     }
 
-    public function testFromReflectionOmitsTypeHintInTypedProperty(): void
+    public function testFromReflectionWithTypeHintInTypedProperty(): void
```
please don't change existing tests to avoid regression, add new one instead.

allansun replied:
This was changed on purpose of adding 'type hinting' to 'TypedProperty'. Without modifying this, the test would definitely fail. Please advise what I should do.

samsonasik replied:
except `readonly` which must has type hint, this new feature should be optional to avoid regression.

samsonasik replied:
You may need to add `$omitTypeHint` 2nd parameter to `PropertyGenerator::fromReflection` method, with default true. 

@Ocramius any opinion about this?

allansun replied:
I double checked  the composer.json and seems the minimum PHP version requirement is 7.4, which has type hint  for property already. Am I right to say we don't have to omit  the generation of type hint for backward compatibility at all? Hence the need for changing the test is necessary, because it seems to me that the original test was purely created because the lack of ability to generate type hint which should have been done a long time ago.

allansun replied:
after making the changes (change $type from string to TypeGenerator)as per requested by @Ocramius 
I think the only conflict we have here is whether we should modify the existing unit test, which was objected by @samsonasik.

Let me re-clarify the reason I changed the existing unit test:
1. The old test naming was a bit misleading. There was no option to 'omit' type hint, because there was no way to generate type hint on property until my PR.
2. The old test was created in https://github.com/laminas/laminas-code/commit/0872dbda5d3a8b680a6af4dae4fe76f7e0c9714a , and in the original PR https://github.com/laminas/laminas-code/pull/77 the author explained clearly: 'The test explicitly checks, that typed properties will be ignored when generating the code. Maybe that feature should be implemented in a future release, and that will cause this test to fail.'

So it's pretty obvious to me such change to the existing unit test is necessary / mandatory.

allansun committed:
Code improvement for psalm checks
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -21,10 +24,38 @@ class PropertyGenerator extends AbstractMemberGenerator
 
     protected bool $isConst = false;
 
+    protected ?string $type = null;
```
Uhhh, let's avoid having even more state :| 

Also, what is a `$type` here? Shouldn't be a `string`, but rather a `TypeGenerator`.

allansun replied:
You are absolutely RIGHT! Silly me, just changed it and pushed!

allansun committed:
PropertyGenerator::$type change from string to TypeGenerator
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

allansun committed:
Add backward compatibility for PropertyGenerator::fromReflection()
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

allansun committed:
Fix failed unit test under PHP 8.1
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

samsonasik commented on test/Generator/PropertyGeneratorTest.php:
```diff
@@ -423,7 +423,7 @@ public function testFromReflectionReadonlyProperty(): void
         $code      = $generator->generate();
 
-        self::assertSame('    public readonly string $readonly;', $code);
+        self::assertSame('    public readonly $readonly;', $code);
```
@allansun readonly is required to have type

allansun replied:
@samsonasik it was not my test. I removed the type as it should be (check code history, I added it there, which you objected in your earlier review). 

For BC / regression purpose, the result is expected to be without type hint

samsonasik replied:
ok 👍 , for note: separate PR is required to solve it as it will produce fatal error https://3v4l.org/n9DA4

allansun replied:
Noooope, I don't think so... 

First of all, I didn't do the readonly part, it was already merged into the branch before I started working on it. So presumably all existing code should have passed all the tests.

Secondly please refer to comment by  @Ocramius https://discourse.laminas.dev/t/rfc-laminas-laminas-code-add-type-hinting-option-in-propertygenerator/2652/5 I think he's making a valid point that it should depend on the developer to choose which version of PHP the generated code should be compatible with.

Thirdly I can't see why the test would fail since it has /** @requires PHP >= 8.1 */ mark on it

https://github.com/allansun/laminas-code/blob/02b822faec472398edff890077b583922edfd264/test/Generator/PropertyGeneratorTest.php#L413-L427

samsonasik replied:
@allansun my note points is not must be you to implement typed readonly, can be other contributor to implement it :)

allansun commented:
Additional switch added to PropertyGenerator:fromReflection() for BC reason.

@samsonasik Your worry should be eased now with the introduction of this switch, I've added back the unit test checking omitted type hint

@Ocramius As per discussion on the RFC topic, I believe all areas should be covered now.

If everybody's happy with this PR can somebody please approve it? As my work was stuck by such feature

allansun commented:
To be honest I didn’t quite get you on the separate PR, what should I do?

On Tue, 14 Dec 2021 at 13:52, Abdul Malik Ikhsan ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/Generator/PropertyGeneratorTest.php
> <https://github.com/laminas/laminas-code/pull/123#discussion_r768685428>:
>
> > @@ -423,7 +423,7 @@ public function testFromReflectionReadonlyProperty(): void
>          $generator = PropertyGenerator::fromReflection($reflectionProperty);
>          $code      = $generator->generate();
>
> -        self::assertSame('    public readonly string $readonly;', $code);
> +        self::assertSame('    public readonly $readonly;', $code);
>
> @allansun <https://github.com/allansun> my note points is not must be you
> to implement typed readonly, can be other contributor to implement it :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/laminas/laminas-code/pull/123#discussion_r768685428>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFAMYKMOYEOEEDNYNSHVTUQ5DZPANCNFSM5J3V77TA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
Best regards, Allan Sun Sent from Gmail Mobile

samsonasik commented:
@allansun I mean, the original code is incorrect for non-typed `readonly`, which unrelated to your PR, which is ok to not touch it.

It need to be solved still in the future, my note is for future reference, when there is contributor that willing to solve it.

allansun commented:
I see! I am quite willing to do it especially when I see there’re loads of
lines not yet covered by the test yet. I’m one of those who’s always
chasing 100% coverage 🤣

With regards to my PR any ideas when we can merge it onto the branch? My
work is still waiting for this feature. I don’t mind temporarily use an
official dev* package but don’t want to publish my forked repo on packagist
only because I need it in urgent.

On Tue, 14 Dec 2021 at 14:18, Abdul Malik Ikhsan ***@***.***>
wrote:

> @allansun <https://github.com/allansun> I mean, the original code is
> incorrect for non-typed readonly, which unrelated to your PR, which is ok
> to not touch it.
>
> It need to be solved still in the future, my note is for future reference,
> when there is contributor that willing to solve it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/laminas/laminas-code/pull/123#issuecomment-993586842>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFAM473VQSCCTYPQ7C2MDUQ5GZ7ANCNFSM5J3V77TA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
Best regards, Allan Sun Sent from Gmail Mobile

allansun commented:
Oh thanks!

On Tue, 14 Dec 2021 at 14:20, Abdul Malik Ikhsan ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/laminas/laminas-code/pull/123#pullrequestreview-831563838>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFAM6G334BPYSE545EEW3UQ5HBTANCNFSM5J3V77TA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
Best regards, Allan Sun Sent from Gmail Mobile

samsonasik commented:
@allansun why you close it?

allansun commented:
I saw you merged it ?

On Tue, 14 Dec 2021 at 15:10, Abdul Malik Ikhsan ***@***.***>
wrote:

> @allansun <https://github.com/allansun> why you close it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/laminas/laminas-code/pull/123#issuecomment-993639562>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFAM64NPRITHHWGAGMB4LUQ5M4TANCNFSM5J3V77TA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
Best regards, Allan Sun Sent from Gmail Mobile

samsonasik commented:
Not yet merged, I just approve it, I think it should wait for @Ocramius to decide about it.

allansun commented:
Sorry I was on the move and didn’t check properly. Reopened it now.

cc @Ocramius

allansun commented:
@Ocramius Could you review the above changes and merge it asap if all approved? My work is in urgent need for it.

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -21,12 +21,40 @@ class PropertyGenerator extends AbstractMemberGenerator
+
     /** @return static */
-    public static function fromReflection(PropertyReflection $reflectionProperty)
+    public static function fromReflection(PropertyReflection $reflectionProperty, bool $enablePropertyTypeHint = false)
```
Boolean flag to be removed: if we're reflecting a symbol, and that symbol already has type declarations, then we're supposed to generate types for it (and also by default, like with all other `::fromReflection()` methods

allansun replied:
Who said we need to be backward compatible? 😂
If I remove this somebody gonna jump out and said it's not BC...

allansun replied:
Let's say some weirdo took a class in PHP 7.4 standard and trying to generate code for, say, PHP 7.1 (I was gonna say 5 😂), then this flag shows it's value.

Ocramius replied:
Not a problem - either you remove it, or I remove it before merging :P

allansun replied:
To remove this flag, I have to modify the unit test @samsonasik [objected before](https://github.com/laminas/laminas-code/pull/123#pullrequestreview-829578913)

If you say so sir, I'll get it removed straight away.

Ocramius replied:
@samsonasik no offense, but let's get rid of this stuff :D

allansun replied:
Just removed the flag as per requested by @Ocramius

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -304,4 +343,14 @@ public function omitDefaultValue(bool $omit = true)
+        return $this->type;
+    }
+
+    public function setType(TypeGenerator $type): void
```
Can the type be somehow reset? If we allow `?TypeGenerator $type` here, then, the constructor can call this method all the time (and a conditional disappears too)

allansun replied:
I simply followed convention like all other properties of PropertyGenerator.

I couldn't think of a need to RESET the $type, that's why I took the '?' out of `?TypeGenerator $type`

Ocramius replied:
> I couldn't think of a need to RESET the `$type`

This class is mostly a builder, so I can imagine `$p = PropertyGenerator::fromReflection(...); $p->setType(null); ...`

Ocramius replied:
49d33430a8d09699aa5aed85bcb0b3bfdf7033c1

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -21,12 +21,40 @@ class PropertyGenerator extends AbstractMemberGenerator
+        if ($flags !== self::FLAG_PUBLIC) {
+            $this->setFlags($flags);
+        }
+        if (null !== $type) {
```
Conditional can be removed, if `setType()` accepts `null` too

allansun replied:
Remind me again, `setDefaultValue()` can take `null`, and there's a check for `null` in this constructor.

Why do we do it differently here?

Ocramius replied:
49d33430a8d09699aa5aed85bcb0b3bfdf7033c1

Ocramius commented on src/Generator/PropertyGenerator.php:
```diff
@@ -136,43 +176,38 @@ public static function fromArray(array $array)
                     $property->setReadonly($value);
                     break;
+                case 'type':
+                    if (! $value instanceof TypeGenerator) {
```
I'm so gonna deprecate all these `::fromArray()` horrors someday :D

allansun replied:
Never used the fromArray nor fromReflection methods anyway 😂

Ocramius reviewed (changes requested):
Further simplification necessary (sorry, haven't checked above discussion, but the whole "make this optional" is really not needed :D )

allansun committed:
Remove $enablePropertyTypeHint flag from PropertyGenerator::fromReflection()
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

allansun committed:
Fix unit test for psalm check
Signed-off-by: Allan Sun <allan.sun@bricre.com>

Signed-off-by: Allan Sun <allan.sun@bricre.com>

nawarian commented:
Hello all! 👋 

I'd benefit a lot from this feature. Wondering if you need any help to get this over the finish line, I'd be happy to get involved if necessary.

Cheers!

Ocramius commented:
I marked this as requiring another review: probably able to work on it after PHPDay 2022.

nawarian commented:
Sounds awesome! Best of luck with the PHP day too!

If you need any support, feel free to ping me.

Cheers!

RafaelKr commented:
Can there anything be done to move this forward?

I would love if this can be used for https://github.com/goetas-webservices/xsd2php

Ocramius commented:
Just picked this up, and found that the generated strings were all wrong anyway :D 

```
1) LaminasTest\Code\Generator\PropertyGeneratorTest::testPropertyCanProduceTypeHinting
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'    public SomeClass $someVal = 'value';'
+'    public \SomeClass $someVal = 'value';'
```

Pushing a few fixes to contributor branch, then merging + shipping :rocket:

Ocramius committed:
Fixed how #123 generates property types, consider nullability and FQCN references too

Before this change, generated property types were relative to current namespace, as well
as missing nullability flag.

This small change ensures that generated types match expectations, and make sense at runtime too.

Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Allow `PropertyGenerator#setType()` to accept `null`, removing therefore the property type

In #123, some discussion around whether a type could be reset occurred: this change allows
for that, so it is possible to generate a `PropertyGenerator` from reflection, and then remove
the property type (if needed).

This also led to reducing a lot of code (removed conditionals).

Ref: https://github.com/laminas/laminas-code/pull/123#discussion_r769513211
Signed-off-by: Marco Pivetta <ocramius@gmail.com>

Ocramius committed:
Adjusted minor static analysis issue caused by non-immutable API

`$this->type` could mutate while being accessed, so we must stash its value in a variable, to prevent potential NPEs

Ocramius commented:
:ship: thanks for the patience, everyone!
